internal/app/orchestrator/orchestrator_test.go

```diff
@@ -44,3 +123,216 @@ func TestNew(t *testing.T) {
+	// We expect a watch to be open but no response.
+	req2 := gcp.Request{
+		VersionInfo: "2",
+		TypeUrl:     "type.googleapis.com/envoy.api.v2.Listener",
+	}
+
+	respChannel2, cancelWatch2 := orchestrator.CreateWatch(req2)
+	assert.NotNil(t, respChannel2)
+	assert.Equal(t, 2, len(orchestrator.downstreamResponseMap.responseChannels))
+	testutils.AssertSyncMapLen(t, 1, orchestrator.upstreamResponseMap.internal)
+
+	// If we pass this point, it's safe to assume the respChannel2 is empty,
+	// otherwise the test would block and not complete.
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+	orchestrator.shutdown(ctx)
+	testutils.AssertSyncMapLen(t, 0, orchestrator.upstreamResponseMap.internal)
+
+	cancelWatch()
+	assert.Equal(t, 1, len(orchestrator.downstreamResponseMap.responseChannels))
+	cancelWatch2()
+	assert.Equal(t, 0, len(orchestrator.downstreamResponseMap.responseChannels))
+}
+
+func TestMultipleWatchesAndUpstreams(t *testing.T) {
+	upstreamResponseChannelLDS := make(chan *v2.DiscoveryResponse)
+	upstreamResponseChannelCDS := make(chan *v2.DiscoveryResponse)
+	mapper := newMockMapper(t)
+	orchestrator := newMockOrchestrator(
+		t,
+		mapper,
+		mockMultiStreamUpstreamClient{
+			ldsResponseChan: upstreamResponseChannelLDS,
+			cdsResponseChan: upstreamResponseChannelCDS,
+			mapper:          mapper,
+			t:               t,
+		},
+	)
+	assert.NotNil(t, orchestrator)
+
+	req1 := gcp.Request{
+		TypeUrl: "type.googleapis.com/envoy.api.v2.Listener",
+		Node: &v2_core.Node{
+			Id: "req1",
+		},
+	}
+	req2 := gcp.Request{
+		TypeUrl: "type.googleapis.com/envoy.api.v2.Listener",
+		Node: &v2_core.Node{
+			Id: "req2",
+		},
+	}
+	req3 := gcp.Request{
+		TypeUrl: "type.googleapis.com/envoy.api.v2.Cluster",
+		Node: &v2_core.Node{
+			Id: "req3",
+		},
+	}
+
+	respChannel1, cancelWatch1 := orchestrator.CreateWatch(req1)
+	assert.NotNil(t, respChannel1)
+	respChannel2, cancelWatch2 := orchestrator.CreateWatch(req2)
+	assert.NotNil(t, respChannel2)
+	respChannel3, cancelWatch3 := orchestrator.CreateWatch(req3)
+	assert.NotNil(t, respChannel3)
+
+	upstreamResponseLDS := v2.DiscoveryResponse{
+		VersionInfo: "1",
+		TypeUrl:     "type.googleapis.com/envoy.api.v2.Listener",
+		Resources: []*any.Any{
+			&anypb.Any{
+				Value: []byte("lds resource"),
+			},
+		},
+	}
+	upstreamResponseCDS := v2.DiscoveryResponse{
+		VersionInfo: "1",
+		TypeUrl:     "type.googleapis.com/envoy.api.v2.Cluster",
+		Resources: []*any.Any{
+			&anypb.Any{
+				Value: []byte("cds resource"),
+			},
+		},
+	}
+
+	upstreamResponseChannelLDS <- &upstreamResponseLDS
+	upstreamResponseChannelCDS <- &upstreamResponseCDS
+
+	gotResponseFromChannel1 := <-respChannel1
+	gotResponseFromChannel2 := <-respChannel2
+	gotResponseFromChannel3 := <-respChannel3
+
+	assert.Equal(t, 3, len(orchestrator.downstreamResponseMap.responseChannels))
+	testutils.AssertSyncMapLen(t, 2, orchestrator.upstreamResponseMap.internal)
```

Comment:
Is asserting the contents of the channel good enough vs actually checking the keys?

Reply:
I primarily cared about the length since I do value checking below when the content flows downstream:
```
assertEqualResources(t, gotResponseFromChannel1, upstreamResponseLDS, req1)
assertEqualResources(t, gotResponseFromChannel2, upstreamResponseLDS, req2)
assertEqualResources(t, gotResponseFromChannel3, upstreamResponseCDS, req3)
```
but I agree that a key/value check will be a more complete test. I'll make the change.